spark/v2.3/spark/src/main/java/com/netease/arctic/spark/source/ArcticSource.java

```diff
@@ -163,4 +178,54 @@ private ArcticCatalog catalog(RuntimeConfig conf) {
     String url = ArcticSparkUtil.catalogUrl(conf);
     return CatalogLoader.load(url);
   }
+
+  @Override
+  public DataSourceReader createReader(DataSourceOptions options) {
+    ArcticTable arcticTable = getTableWithPath(options);
+    SparkSession spark = SparkSession.getActiveSession().get();
+    if (arcticTable.isKeyedTable()) {
+      return new ArcticKeyedTableScan(spark, arcticTable.asKeyedTable());
+    } else {
+      return new ArcticUnkeyedTableScan(spark, arcticTable.asUnkeyedTable());
+    }
+  }
+
+  @Override
+  public Optional<DataSourceWriter> createWriter(String jobId, StructType schema,
+                                                 SaveMode mode, DataSourceOptions options) {
+    ArcticTable arcticTable = getTableWithPath(options);
+    if (arcticTable.isKeyedTable()) {
+      if (mode == SaveMode.Overwrite) {
+        return Optional.of(new ArcticKeyedSparkOverwriteWriter(arcticTable.asKeyedTable(), schema, options));
+      } else if (mode == SaveMode.Append) {
+        // TODO: support keyed append
+        throw new UnsupportedOperationException("Not support now!");
+      }
+    } else if (arcticTable.isUnkeyedTable()) {
+      if (mode == SaveMode.Overwrite) {
+        return Optional.of(new ArcticUnkeyedSparkOverwriteWriter(arcticTable.asUnkeyedTable(), schema, options));
+      } else if (mode == SaveMode.Append) {
+        // TODO: support unkeyed append
+        throw new UnsupportedOperationException("Not support now!");
+      }
+    } else {
+      throw new UnsupportedOperationException("Illegal table type!");
+    }
+    return Optional.empty();
+  }
```

Comment:
we shoud support create table as select.  so in this place, check table exists first, if not , create table and then return writer.

you can return writer from table.createWrite